### PR TITLE
Fix for issue #5852: _translateTime function in I18n returns a (too deep...

### PR DESCRIPTION
**Pull request by mvdriel**
... nested) array

### CODE REVIEW

**Comment by dereuromark** [comment]
A testcase would be great to avoid future regression.

**Comment by markstory** [comment]
Or even a way to reproduce the issue.

**Comment by tersmitten** [comment]
I will add some more details on monday. The only details I can add now is that warnings were triggered from CakeTime and that we are using translations from app/Locale/*/LC_TIME/

**Comment by mvdriel** [comment]
It seems that po-files for LC_TIME are not supported anymore since the introduction of context.

I now see that my fix is only correct for LC_TIME using po-files and not for 'normal' LC_TIME-files, because context is missing for the latter one. It is probably better to fix this by removing the context level (in the translation strings array $_domains) also for LC_TIME po-files.

**Comment by tersmitten** [comment]
Thanks @mvdriel for adding more details

**Comment by markstory** [comment]
When you say lc_time po files. Are you using standard po files in place of the specific lc_time file?

**Comment by mvdriel** [comment]
@markstory That's right.

**Comment by markstory** [comment]
@mvdriel Why are you doing that? I don't think having that work was ever intended to work.

**Comment by mvdriel** [comment]
@markstory Before CakePHP 2.6.x the Cake I18n extract did extract those strings to /app/Locale/LC_TIME/cake.pot, so you were able to translate them customized. How is this customizing possible when using system locales or a project-based lc_time file?

**Comment by markstory** [comment]
You would have to use the lc_time formatted files to customize the messages.

**Comment by mvdriel** [comment]
@markstory Is there an easy way to customize lc_time formatted files? They look a bit "strange". :)

**Comment by ADmad** [comment]
@mvdriel LC_TIME format is standard for each language. You can find the LC_TIME files for common languages in the [localized](https://github.com/cakephp/localized/tree/master/Locale) repo.

**Comment by tersmitten** [comment]
@markstory, but editing lc_time files is tedious to do:

```
LC_TIME
abday   "<U007A><U006F>";"<U006D><U0061>";"<U0064><U0069>";/
	"<U0077><U006F>";"<U0064><U006F>";"<U0076><U0072>";/
	"<U007A><U0061>"
day     "<U007A><U006F><U006E><U0064><U0061><U0067>";/
	"<U006D><U0061><U0061><U006E><U0064><U0061><U0067>";/
	"<U0064><U0069><U006E><U0073><U0064><U0061><U0067>";/
	"<U0077><U006F><U0065><U006E><U0073><U0064><U0061><U0067>";/
	"<U0064><U006F><U006E><U0064><U0065><U0072><U0064><U0061><U0067>";/
	"<U0076><U0072><U0069><U006A><U0064><U0061><U0067>";/
	"<U007A><U0061><U0074><U0065><U0072><U0064><U0061><U0067>"
abmon   "<U006A><U0061><U006E>";"<U0066><U0065><U0062>";/
	"<U006D><U0072><U0074>";"<U0061><U0070><U0072>";/
	"<U006D><U0065><U0069>";"<U006A><U0075><U006E>";/
	"<U006A><U0075><U006C>";"<U0061><U0075><U0067>";/
	"<U0073><U0065><U0070>";"<U006F><U006B><U0074>";/
	"<U006E><U006F><U0076>";"<U0064><U0065><U0063>"
mon     "<U006A><U0061><U006E><U0075><U0061><U0072><U0069>";/
	"<U0066><U0065><U0062><U0072><U0075><U0061><U0072><U0069>";/
	"<U006D><U0061><U0061><U0072><U0074>";/
	"<U0061><U0070><U0072><U0069><U006C>";/
	"<U006D><U0065><U0069>";/
	"<U006A><U0075><U006E><U0069>";/
	"<U006A><U0075><U006C><U0069>";/
	"<U0061><U0075><U0067><U0075><U0073><U0074><U0075><U0073>";/
	"<U0073><U0065><U0070><U0074><U0065><U006D><U0062><U0065><U0072>";/
	"<U006F><U006B><U0074><U006F><U0062><U0065><U0072>";/
	"<U006E><U006F><U0076><U0065><U006D><U0062><U0065><U0072>";/
	"<U0064><U0065><U0063><U0065><U006D><U0062><U0065><U0072>"
d_t_fmt "<U0025><U0061><U0020><U0025><U0064><U0020><U0025><U0062><U0020><U0025><U0059><U0020><U0025><U0054><U0020><U0025><U005A>"
d_fmt   "<U0025><U0064><U002D><U0025><U006D><U002D><U0025><U0079>"
t_fmt   "<U0025><U0054>"
am_pm   "";""
t_fmt_ampm ""
date_fmt       "<U0025><U0061><U0020><U0025><U0062><U0020><U0025><U0065>/
<U0020><U0025><U0048><U003A><U0025><U004D><U003A><U0025><U0053><U0020>/
<U0025><U005A><U0020><U0025><U0059>"

week    7;19971130;4
first_weekday 2
first_workday 2
END LC_TIME
```

Working with `app/Locale/*/LC_TIME/cake.po` is much easier :-)

**Comment by ADmad** [comment]
@tersmitten Why would you edit a LC_TIME file?  It already has standard formats for particular language.

**Comment by tersmitten** [comment]
@ADmad Because `d_fmt   "%d-%m-%y"` does not make sense in Dutch for instance (and there are more examples)

**Comment by ADmad** [comment]
@tersmitten In that case you should contact the relevant maintainers of locale files in linux and find out why they think it does :)

To the best of my knowledge the LC_TIME files in the localized repo have been taken from locale files in linux for each language. If you think the file for a particular language is incorrect please submit a PR with correct one.

File: lib/Cake/I18n/I18n.php [hunk 1]
```diff
@@ -727,14 +727,20 @@ protected function _parseLiteralValue($string) {
...
  */
-	protected function _translateTime($format, $domain) {
-		if (!empty($this->_domains[$domain][$this->_lang]['LC_TIME'][$format])) {
-			if (($trans = $this->_domains[$domain][$this->_lang][$this->category][$format])) {
-				return $trans;
-			}
+	protected function _translateTime($format, $domain, $context) {
+		$lang = $this->_lang;
+		$category = $this->category;
+		$trans = Hash::get($this->_domains, "$domain.$lang.$category.$format.$context");
```
Comment (markstory) [inline lib/Cake/I18n/I18n.php]: When would there ever be a context for LC_TIME?

**Comment by markstory** [comment]
I can understand that LC_TIME files are a bit of a pain, but they are a somewhat standard file used in various systems. Having two ways to translate time data seems a bit concerning.

File: lib/Cake/I18n/I18n.php (hunk 1 above)
Comment (mvdriel) [inline lib/Cake/I18n/I18n.php]: I suppose for lc_time context will always be an empty string, but it is needed to fetch the translation for lc_time po files from the correct index in the translations array ($_domains).

Comment (markstory) [inline lib/Cake/I18n/I18n.php]: Yeah, instead of a parameter, '' could be hardcoded.

**Comment by tersmitten** [comment]
We created https://github.com/Oefenweb/glibc-locale-tools, that should make it easier for us to handle `LC_TIME` files. Still I liked the `.po` files for `LC_TIME`.

**Comment by markstory** [comment]
Closing as we're not going to continue support `.po` files for LC_TIME files. On the upside LC_TIME files are going away in 3.0.

**Comment by tersmitten** [comment]
Sure, we already fixed the problem and successfully upgraded to 2.6.2

**Pull request closed without merging**